server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1313,6 +1313,13 @@ public void close(String reason, boolean flushEngine) throws IOException {
         }
     }
 
+    public void preRecovery() {
+        if (state != IndexShardState.RECOVERING) {
```

Comment:
More for my curiosity than something that needs changing: why can we not `assert state == IndexShardState.RECOVERING` here? The lifecycle of an `IndexShard` is still a bit opaque to me.

Reply:
I think it makes more sense to assert here, there's no reason to not do it.